subsys/bluetooth/services/ots/ots.c

```diff
@@ -271,63 +288,86 @@ static ssize_t ots_obj_prop_read(struct bt_conn *conn,
 				 sizeof(ots->cur_obj->metadata.props));
 }
 
-int bt_ots_obj_add(struct bt_ots *ots,
-			    struct bt_ots_obj_metadata *obj_init)
+int bt_ots_obj_add_internal(struct bt_ots *ots, struct bt_conn *conn, uint32_t size,
+			    const struct bt_ots_obj_type *type, struct bt_gatt_ots_object **obj)
 {
 	int err;
-	struct bt_gatt_ots_object *obj;
-	size_t name_len;
+	struct bt_gatt_ots_object *new_obj;
 
 	if (IS_ENABLED(CONFIG_BT_OTS_DIR_LIST_OBJ) && ots->dir_list &&
 	    ots->dir_list->dir_list_obj->state.type != BT_GATT_OTS_OBJECT_IDLE_STATE) {
 		LOG_DBG("Directory Listing Object is being read");
 		return -EBUSY;
 	}
 
-	name_len = strlen(obj_init->name);
-
-	CHECKIF(name_len == 0 || name_len > CONFIG_BT_OTS_OBJ_MAX_NAME_LEN) {
-		LOG_DBG("Invalid name length %zu", name_len);
-		return -EINVAL;
-	}
-
-	CHECKIF(!ots_obj_validate_prop_against_oacp(obj_init->props, ots->features.oacp)) {
-		LOG_DBG("Object properties (0x%04X) are not a subset of OACP (0x%04X)",
-				obj_init->props, ots->features.oacp);
-		return -ENOTSUP;
-	}
-
-	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &obj);
+	err = bt_gatt_ots_obj_manager_obj_add(ots->obj_manager, &new_obj);
```

Comment:
We are we attempting to add the object to the object manager, before we ask the upper layer for the values such as the size?

Reply:
That was the original order of operations here but you are correct, it seems odd. I will look into changing it.

Reply:
After looking at this, I think this is the appropriate order of operations. The intention is to ensure the `obj_manager` is capable of holding the new object before bothering the application layer.

I would assume that the creation operation in the application layer is fairly expensive (in my application it requires writing the object to a SPI FLASH device) and this order of operation prevents a situation where we have to delete a created object because the `obj_manager` has no room left.

Reply:
> After looking at this, I think this is the appropriate order of operations. The intention is to ensure the obj_manager is capable of holding the new object before bothering the application layer.

That makes sense, but perhaps we could query the object manager to see if there's 1 more spot, so we don't need to add and remove. This may be a bit tricky to make thread safe though.

Reply:
I took a closer look at the object manager, I do not believe it is actually thread safe. Am I misunderstanding something? I presume the intention is that the object manager is only called from the Bluetooth receive thread and as such doesn't need to thread safe?

If thread safety is not needed then your suggestion makes sense and I can change this to what you have suggested.

Reply:
Agreed, the object manager is actually not threadsafe in its current implementation, as it may be updated from the RX thread or the application thread(s).

Reply:
@Thalley, there is actually another hitch in this plan, the `obj_created` callback takes the ID of the created object. So it is more than just available room, the object manager needs to give us an ID for this object.

We could add a reservation mechanism for the object manager to hold an object for us but at that point is it really that different from adding the object and then later removing it? I presume if we supported the concurrency feature and multiple connections then we could distinguish between reservations and actual object additions where reserved objects were not visible to a 2nd connection until the create procedure had fully completed.

Having said that we could optimize this a bit and not add the object to the directory list object until we know for sure the application has created the object.

What do you think?

Reply:
> Having said that we could optimize this a bit and not add the object to the directory list object until we know for sure the application has created the object.

Isn't that already the case? 

But I agree with your comment, and that we can't really do it much different than this.